### DISCUSSION TITLE
Nats request and reply is blocking the other requerst in pipeline

### DISCUSSION THREAD

**Discussion by khanakia** [nats-io/nats.go#1541] [Q&A]
If i send both requests one after another, the second request has to wait for the first request to be completed. I do not want the second request to be wait it should work more like an HTTP request-reply pattern.

However, I can put the func inside `go func {}()` but I am not sure if I am doing something wrong or if nats has some built in solution

Note: `test 5` means delay the response for the 5 seconds dynamically

```
nats request test 5
nats request test 0
```

```go
nats.GetEncodedConn().Subscribe(Subject("test"), func(subj, reply string, msg int) {
  time.Sleep(time.Duration(msg) * time.Second)
  ec.Publish(reply, "done")
})
```

**Comment by wallyqs** [comment]
Hi @khanakia this is how the clients are currently implemented, head of line blocking per request by default and in order to increase concurrency you would wrap this under a goroutine as you point out.

**Comment by derekcollison** [comment]
You can move to async processing with PublishReply IIRC.

**Reply by khanakia** [reply]
But i need it to be synchronous (request-reply) and get the reply immediately. As we have Rest API Gateway and all the endpoint uses NATS to send requests to Microservices and get responses so we can send the same to the Web Browser Frontend
